utils/s2n_map.c

```diff
@@ -0,0 +1,210 @@
+    struct s2n_map_entry *table;
+};
+
+static uint32_t s2n_map_slot(struct s2n_map *map, struct s2n_blob *key)
+{
+    struct s2n_hash_state sha256;
+    union {
+        uint8_t u8[32];
+        uint32_t u32[8];
+    } digest;
+
+    GUARD(s2n_hash_init(&sha256, S2N_HASH_SHA256)); 
+    GUARD(s2n_hash_update(&sha256, key->data, key->size)); 
+    GUARD(s2n_hash_digest(&sha256, digest.u8, sizeof(digest)));    
+
+    return digest.u32[0] % map->capacity;
+}
+
+static int s2n_map_embiggen(struct s2n_map *map, uint32_t capacity)
+{
+    struct s2n_blob mem;
+    struct s2n_map tmp; 
+
+    if (map->immutable) {
+        S2N_ERROR(S2N_ERR_MAP_IMMUTABLE);
+    }
+
+    GUARD(s2n_alloc(&mem, (capacity * sizeof(struct s2n_map_entry))));
+
+    tmp.capacity = capacity;
+    tmp.size = 0;
+    tmp.table = (void *) mem.data;
+    tmp.immutable = 0;
+
+    for (int i = 0; i < map->capacity; i++) {
+        if (map->table[i].key.size) {
+            GUARD(s2n_map_add(&tmp, &map->table[i].key, &map->table[i].value));
+            GUARD(s2n_free(&map->table[i].key));
+            GUARD(s2n_free(&map->table[i].value));
+        }
+    }
+
+    /* Free the old memory */
+    mem.data = (void *) map->table;
+    mem.size = map->capacity * sizeof(struct s2n_map_entry);
+    GUARD(s2n_free(&mem));
+
+    /* Clone the temporary map */
+    map->capacity = tmp.capacity;
+    map->size = tmp.size;
+    map->table = tmp.table;
+    map->immutable = 0;
+
+    return 0;
+}
+
+struct s2n_map *s2n_map_new()
+{
+    struct s2n_blob mem; 
+    struct s2n_map *map;
+   
+    GUARD_PTR(s2n_alloc(&mem, sizeof(struct s2n_map)));
+
+    map = (void *) mem.data;
+    map->capacity = 0;
+    map->immutable = 0;
+    map->table = NULL;
+
+    GUARD_PTR(s2n_map_embiggen(map, S2N_INITIAL_TABLE_SIZE));
+    
+    return map;
+}
+
+int s2n_map_add(struct s2n_map *map, struct s2n_blob *key, struct s2n_blob *value)
+{
+    if (map->immutable) {
+        S2N_ERROR(S2N_ERR_MAP_IMMUTABLE);
+    }
+
+    if (map->capacity < (map->size * 2)) {
+        /* Embiggen the map */
+        GUARD(s2n_map_embiggen(map, map->capacity * 2));
+    }
+
+    uint32_t slot = s2n_map_slot(map, key);
+    
+    /* Linear probing until we find an empty slot */
+    while(map->table[slot].key.size) {
+        if (key->size != map->table[slot].key.size ||
+            memcmp(key->data,  map->table[slot].key.data, key->size)) {
+            slot++;
+            slot %= map->capacity; 
+            continue;
+        }
+
+        /* We found a duplicate key */
+        S2N_ERROR(S2N_ERR_MAP_DUPLICATE);
+    }
+    
+    GUARD(s2n_dup(key, &map->table[slot].key));
+    GUARD(s2n_dup(value, &map->table[slot].value));
+    map->size++;
+
+    return 0;
+}
+
+int s2n_map_complete(struct s2n_map *map)
+{
+    map->immutable = 1;
+
+    return 0;
+}
+
+int s2n_map_lookup(struct s2n_map *map, struct s2n_blob *key, struct s2n_blob *value)
+{
+    if (!map->immutable) {
+        S2N_ERROR(S2N_ERR_MAP_IMMUTABLE);
```

Comment:
This error message references "updating" the map and is confusing in the context of a lookup operation. Perhaps the message could be more generally worded to accommodate both cases.